src/Bicep.LangServer/Handlers/BicepDocumentFormattingHandler.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Threading;
+using System.Threading.Tasks;
+using Bicep.Core.PrettyPrint;
+using Bicep.LanguageServer.CompilationManager;
+using Bicep.LanguageServer.Utils;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Bicep.Core.PrettyPrint.Options;
+using Microsoft.Extensions.Logging;
+using Bicep.Core.Syntax;
+using Bicep.LanguageServer.Extensions;
+
+namespace Bicep.LanguageServer.Handlers
+{
+    public class BicepDocumentFormattingHandler : DocumentFormattingHandler
+    {
+        private readonly ILogger<BicepDocumentSymbolHandler> logger;
+        private readonly ICompilationManager compilationManager;
+
+        public BicepDocumentFormattingHandler(ILogger<BicepDocumentSymbolHandler> logger, ICompilationManager compilationManager)
+            : base(CreateRegistrationOptions())
+        {
+            this.logger = logger;
+            this.compilationManager = compilationManager;
+        }
+
+        public override Task<TextEditContainer?> Handle(DocumentFormattingParams request, CancellationToken cancellationToken)
+        {
+            CompilationContext? context = this.compilationManager.GetCompilation(request.TextDocument.Uri);
+
+            if (context == null)
+            {
+                // we have not yet compiled this document, which shouldn't really happen
+                this.logger.LogError("Document formatting request arrived before file {Uri} could be compiled.", request.TextDocument.Uri);
+
+                return Task.FromResult<TextEditContainer?>(null);
+            }
+
+            long indentSize = request.Options.TabSize;

```

Comment:
Is there any ability in VSCode to set a default spacing for the language? If possible, I'd like the majority of our samples to be formatted with indent = 2, indentkind = space.

Reply:
Yes, the following settings should do the work:
```
    "[bicep]": {
        "editor.tabSize": 2,
        "editor.insertSpaces": true
    }
```

Reply:
Nice!

Reply:
Is that specified in an individual user's settings or picked up as a default when the extension is in use?

Reply:
Agree with what anthony is proposing - it would be good if this can be the default setting for the extension

Reply:
+1 on this as well.

Reply:
It would be up to the users to provide the settings. I also agree that if we can provide some default formatting settings so users can get a consistent experience. My question is should we do that via some custom VSCode bicep settings, or we read it from some external config file, say a .editorconfig file (or a bicep project file if we are going to support that)?

Reply:
Sorry I was wrong. It turns out we can add the following section to the extension manifest in `package.json` - no need to invent custom VSCode settings:

```
    "configurationDefaults": {
      "[bicep]": {
        "editor.tabSize": 2,
        "editor.insertSpaces": true
      }
    },

```

Reply:
That's a great find!

Reply:
I've just done tests with editorconfig as well. As long as the editorconfig VSCode extension is installed, it will override the `tabSize` and `insertSpaces` settings temporarily for the current opened workspace, which means we don't even need to read editorconfig files explicitly by ourself, so we can close this one :)